app/jobs/appointment_notification/worker.rb

```diff
@@ -6,50 +6,77 @@ class AppointmentNotification::Worker
 
   DEFAULT_LOCALE = :en
 
-  def perform(appointment_id, communication_type, locale = nil)
-    appointment = Appointment.find_by(id: appointment_id)
-    unless appointment
-      logger.warn "Appointment #{appointment_id} not found, skipping notification"
+  def perform(appointment_reminder_id, communication_type)
+    reminder = AppointmentReminder.includes(:appointment, :patient).find(appointment_reminder_id)
+    return if reminder.appointment.previously_communicated_via?(communication_type)
+    if reminder.status != "scheduled"
+      report_error("scheduled appointment reminder has invalid status")
       return
     end
+    send_message(reminder, communication_type)
+  end
 
-    return if appointment.previously_communicated_via?(communication_type)
+  private
 
-    patient_phone_number = appointment.patient.latest_mobile_number
-    message = appointment_message(appointment, communication_type, locale)
+  def send_message(reminder, communication_type)
+    notification_service = NotificationService.new
 
     begin
-      notification_service = NotificationService.new
-
       response = if communication_type == "missed_visit_whatsapp_reminder"
-        notification_service.send_whatsapp(patient_phone_number, message, callback_url)
+        notification_service.send_whatsapp(
+          phone_number(reminder.patient),
+          appointment_message(reminder),
+          callback_url
+        )
       else
-        notification_service.send_sms(patient_phone_number, message, callback_url)
+        notification_service.send_sms(
+          phone_number(reminder.patient),
+          appointment_message(reminder),
+          callback_url
+        )
       end
-
-      Communication.create_with_twilio_details!(
-        appointment: appointment,
-        twilio_sid: response.sid,
-        twilio_msg_status: response.status,
-        communication_type: communication_type
-      )
+      create_communication(reminder, communication_type, response)
+      mark_reminder_sent(reminder)
```

Comment:
Extracting a method here to send a message directly to the object feels unnecessary.